// components/BreadCrum/BreadCrum.js
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const uniqueId = () => Math.random();

const BreadcrumContainer = styled.p`
  width: 100%;
  margin: 10px 0px;
  text-align: left;

  i {
    color: #4E4E4E;
    line-height: 9px;
    position: absolute;
    margin-top: 7px;
    margin-left: 5px;
    font-size: 6px;
  }
`;

const BreadcrumItemLast = styled.span`
  color: #000;
  color: #6090C3;
  cursor: pointer;
  font-size: 11px;
  margin-right: 12px;
`;

const BreadcrumItem = styled.span`
  color: #6090C3;
  cursor: pointer;
  font-size: 11px;
  margin-right: 15px;
`;

export default class BreadCrum extends Component {

  static propTypes = {
    style: PropTypes.object,
    hierarchy: PropTypes.array.isRequired,
    handleBreadCrumClick: PropTypes.func
  }

  getBreadCrums() {
    const {hierarchy, handleBreadCrumClick} = this.props;
    const breadcrums = [];
    let count = 1;
    hierarchy.forEach((item) => {
      if (item) {
        const currHierarchy = hierarchy.slice(0, count);
        let lastitem = false;
        let icon = <i>></i>;
        const style = this.props.style ? this.props.style : {};
        if (count === hierarchy.length) {
          icon = <i></i>;
          lastitem = true;
        }
        if (!handleBreadCrumClick) {
          style.cursor = 'auto';
        }
        if (item.href && item.href !== null) {
          const id = item.name.toLowerCase().split(' ').join('_');
          const breadcrumitem = (<BreadcrumItem
            id={id}
            onClick={this.handleBreadCrumClick.bind(this, currHierarchy)}
          >
            {item.name} {icon}
          </BreadcrumItem>);
          const breadcrumitemlast = (<BreadcrumItemLast
            id={id}
            onClick={this.handleBreadCrumClick.bind(this, currHierarchy)}
          >
            {item.name} {icon}
          </BreadcrumItemLast>);

          breadcrums.push(
            <a key={uniqueId()} to={item.href} onClick={item.reload && this.handleReload.bind(this)}>
              {lastitem ? breadcrumitemlast : breadcrumitem}
            </a>
          );
        } else {
          // style.color = '#000';
          breadcrums.push(
            <BreadcrumItem
              key={uniqueId()}
              style={style}
              id={item.name ? item.name.toLowerCase().split(' ').join('_') : ''}
              onClick={this.handleBreadCrumClick.bind(this, currHierarchy)}
            >
              {item.name} {icon}
            </BreadcrumItem>
          );
        }
      }
      count++;
    });
    return breadcrums;
  }

  handleReload() {
    setTimeout(() => {
      window.location.reload();
    }, 2000);
  }

  handleBreadCrumClick(hierarchy) {
    const {handleBreadCrumClick} = this.props;
    if (handleBreadCrumClick) {
      handleBreadCrumClick(hierarchy);
    }
  }

  render() {
    const breadcrums = this.getBreadCrums();
    return (
      <BreadcrumContainer>
        {breadcrums.length > 1 && this.getBreadCrums()}
      </BreadcrumContainer>
    );
  }
}
